### PR TITLE
[resharding] ChatGPT modulation

### PR DESCRIPTION
**Pull request by shreyan-gupta**
Some beautification courtesy ChatGPT.

I double checked everything, we aren't changing any meaning.

### CODE REVIEW

File: neps/nep-0568.md [hunk 1]
```diff
@@ -10,324 +10,244 @@ Created: 2024-10-24
...
+To resolve this, with the new protocol version, we changed the implementation of the runtime to discard executing delayed receipts that don't belong to the `account_id` on that shard.
 
-Note that no delayed receipts are lost during resharding as all receipts get executed exactly once based on which of the child shards does the associated account_id belong to.
+Note that no delayed receipts are lost during resharding, as all receipts get executed exactly once based on which of the child shards the associated `account_id` belongs to.
 
 ### PromiseYield Receipt Handling
 
-Promise Yield were introduced as part of NEP-519 to enable defer replying to caller function while response is being prepared. As part of Promise Yield implementation, it introduced three new trie keys, `PromiseYieldIndices`, `PromiseYieldTimeout` and `PromiseYieldReceipt`.
+Promise Yield was introduced as part of NEP-519 to enable deferring replies to the caller function while the response is being prepared. As part of the Promise Yield implementation, it introduced three new trie keys: `PromiseYieldIndices`, `PromiseYieldTimeout`, and `PromiseYieldReceipt`.
```
Comment (Trisfald) [inline neps/nep-0568.md:188]: is `NEP-519` the thing who introduced the new trie keys?

File: neps/nep-0568.md [hunk 2]
```diff
@@ -10,324 +10,244 @@ Created: 2024-10-24
...
+* `PromiseYieldIndices`: This is a singleton key that holds the `start_index` and `end_index` of the keys in `PromiseYieldTimeout`.
+* `PromiseYieldTimeout { index }`: Along with the `receiver_id` and `data_id`, this stores the `expires_at` block height until which we need to wait to receive a response.
 * `PromiseYieldReceipt { receiver_id, data_id }`: This is the receipt created by the account.
 
 An account can call the `promise_yield_create` host function that increments the `PromiseYieldIndices` along with adding a new entry into the `PromiseYieldTimeout` and `PromiseYieldReceipt`.
 
-The `PromiseYieldTimeout` is sorted as per the time of creation and has an increasing value of expires_at block height. In the runtime, we iterate over all the expired receipts and create a blank receipt to resolve the entry in `PromiseYieldReceipt`.
+The `PromiseYieldTimeout` is sorted by time of creation and has an increasing value of `expires_at` block height. In the runtime, we iterate over all the expired receipts and create a blank receipt to resolve the entry in `PromiseYieldReceipt`.
```
Comment (Trisfald) [inline neps/nep-0568.md:196]: I guess all timeouts are sorted among each others?

File: neps/nep-0568.md [hunk 3]
```diff
@@ -343,165 +263,150 @@ pub struct CongestionInfoV1 {
...
+This field is more complex as it includes information from both delayed receipts and buffered receipts. To calculate this field accurately, we need to know the distribution of `receipt_bytes` across both delayed receipts and buffered receipts. The current solution is to store metadata about the total `receipt_bytes` for buffered receipts in the trie. This way, we have the following:
 
-* For child with lower index, receipt_bytes is the sum of both delayed receipts bytes and congestion control bytes, hence `receipt_bytes = parent.receipt_bytes`
-* For child with upper index, receipt_bytes is just the bytes from delayed receipts, hence `receipt_bytes = parent.receipt_bytes - buffered_receipt_bytes`
+* For the child with the lower index, `receipt_bytes` is the sum of both delayed receipts bytes and buffered receipts bytes, hence `receipt_bytes = parent.receipt_bytes`.
+* For the child with the higher index, `receipt_bytes` is just the bytes from delayed receipts, hence `receipt_bytes = parent.receipt_bytes - buffered_receipt_bytes`.
```
Comment (Trisfald) [inline neps/nep-0568.md:281]: is it more accurate to say `parent.buffered_receipt_bytes`?

File: neps/nep-0568.md [hunk 4]
```diff
@@ -343,165 +263,150 @@ pub struct CongestionInfoV1 {
...
 
-While Frozen MemTries provide the benefits of being compatible with instant resharding, they come at the cost of memory consumption. Once a MemTrie is frozen, since it doesn't support deallocation of memory, it continues to consume as much memory as it did at the time of freezing. In case a node is tracking only one of the child shards, a Frozen MemTrie would continue to use the same amount of memory as the parent trie. Due to this, Hybrid MemTries are only a temporary solution and we rebuild the MemTrie for the children once the post-processing step for Flat Storage is completed.
+While Frozen MemTries facilitate instant resharding, they come at the cost of memory consumption. Once a MemTrie is frozen, it continues to consume the same amount of memory as it did at the time of freezing, as it does not support memory deallocation. If a node tracks only one of the child shards, a Frozen MemTrie would continue to use the same amount of memory as the parent trie. Therefore, Hybrid MemTries are only a temporary solution, and we rebuild the MemTrie for the children once the post-processing step for Flat Storage is completed.
```
Comment (Trisfald) [inline neps/nep-0568.md:354]: `we rebuild the MemTrie for the children once the post-processing step for Flat Storage is completed.`
not sure if we are going to do it in the end, is it worth to put this sentence in the NEP?

File: neps/nep-0568.md [hunk 5]
```diff
@@ -343,165 +263,150 @@ pub struct CongestionInfoV1 {
...
-5. After the new child MemTrie has caught up to the latest block, we do an in-place swap in Client and discard the Hybrid MemTrie.
+1. **Freeze the Parent MemTrie**: Create a read-only frozen arena representing a snapshot of the state at the time of freezing (after post-processing the last block of the epoch). The parent MemTrie is no longer required in runtime going forward.
+2. **Clone the Frozen MemTrie**: Clone the Frozen MemTrie cheaply for both child MemTries to use. This does not clone the parent arena's memory but merely increases the reference count.
+3. **Create Hybrid MemTries for Each Child**: Create a new MemTrie with `HybridArena` for each child. The base of the MemTrie is the read-only `FrozenArena`, while all new node allocations occur in a dedicated `STArena` memory pool for each child MemTrie. This temporary MemTrie is used while Flat Storage is being built in the background.
+4. **Rebuild MemTrie from Flat Storage**: Once the Flat Storage is constructed in the post-processing step of resharding, we use it to load a new MemTrie and catch up to the latest block.
```
Comment (Trisfald) [inline neps/nep-0568.md:363]: same point as line 354

File: neps/nep-0568.md [hunk 6]
```diff
@@ -622,67 +517,65 @@ fn set_shard_uid_mapping(&mut self, child_shard_uid: ShardUId, parent_shard_uid:
...
+The state sync algorithm defines a `sync_hash` used in many parts of the implementation. This is always the first block of the current epoch, which the node should be aware of once it has synced headers to the current point in the chain. A node performing state sync first makes a request (currently to centralized storage on GCS, but in the future to other nodes in the network) for a `ShardStateSyncResponseHeader` corresponding to that `sync_hash` and the Shard ID of the shard it's interested in. Among other things, this header includes the last new chunk before `sync_hash` in the shard and a `StateRootNode` with a hash equal to that chunk's `prev_state_root` field. Then the node downloads (again from GCS, but in the future from other nodes) the nodes of the trie with that `StateRootNode` as its root. Afterwards, it applies new chunks in the shard until it's caught up.
```
Comment (Trisfald) [inline neps/nep-0568.md:526]: nit: we could omit the details about GCS and decentralized state sync, I don't think it has any relevance in this context

**Review by Trisfald — APPROVED** [review]
Overall 👍 

Left a few minor comments

File: neps/nep-0568.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (shreyan-gupta) [reply neps/nep-0568.md:188]: That's grammatically correct. We usually have phrases like "a, b, and c are..."

Comment (shreyan-gupta) [reply neps/nep-0568.md:188]: Yes, promise yield was part of NEP 519

File: neps/nep-0568.md (hunk 4 above)
Comment (shreyan-gupta) [reply neps/nep-0568.md:354]: This is a part of the implementation section but yeah, probably too much information. Let me change it to "we rebuild the MemTrie for the children after resharding is completed".

File: neps/nep-0568.md (hunk 3 above)
Comment (shreyan-gupta) [reply neps/nep-0568.md:281]: Yes that's right. Thanks!

**Pull request merged**